paimon-core/src/main/java/org/apache/paimon/manifest/ManifestFileMeta.java

```diff
@@ -283,16 +283,11 @@ public static Optional<List<ManifestFileMeta>> tryFullCompaction(
                 Predicate predicate = predicateOpt.get();
                 FieldStatsArraySerializer fieldStatsArraySerializer =
                         new FieldStatsArraySerializer(partitionType);
-                for (; j < base.size(); j++) {
-                    // TODO: optimize this to binary search.
-                    ManifestFileMeta file = base.get(j);
-                    if (predicate.test(
-                            file.numAddedFiles + file.numDeletedFiles,
-                            file.partitionStats.fields(fieldStatsArraySerializer))) {
-                        break;
-                    } else {
-                        result.add(file);
-                    }
+                int searchResult =
+                        binarySearchWithFilter(
+                                base, j, base.size() - 1, predicate, fieldStatsArraySerializer);
```

Comment:
I don't remember the specific logic here, but the current modifications are still very dangerous, for example, `j` has not been added here.